FirebaseMLModelDownloader/Sources/TelemetryLogger.swift

```diff
@@ -0,0 +1,163 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import Foundation
+import FirebaseCore
+import GoogleDataTransport
+
+/// Extension to set Firebase app info.
+extension SystemInfo {
+  mutating func setAppInfo(apiKey: String?, projectID: String?) {
+    appID = Bundle.main.bundleIdentifier ?? "unknownBundleID"
+    // appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknownAppVersion"
+    appVersion = "1"
+    self.apiKey = apiKey ?? "unknownAPIKey"
+    firebaseProjectID = projectID ?? "unknownProjectID"
+  }
+}
+
+/// Extension to set model options.
+extension ModelOptions {
+  mutating func setModelOptions(model: CustomModel, isModelUpdateEnabled: Bool? = nil) {
+    if let updateEnabled = isModelUpdateEnabled {
+      self.isModelUpdateEnabled = updateEnabled
+    }
+    modelInfo.name = model.name
+    modelInfo.hash = model.hash
+    modelInfo.modelType = .custom
+  }
+}
+
+/// Extension to build model download log event.
+extension ModelDownloadLogEvent {
+  mutating func setEvent(status: DownloadStatus, errorCode: ErrorCode? = nil,
+                         roughDownloadDuration: UInt64? = nil, exactDownloadDuration: UInt64? = nil,
+                         downloadFailureStatus: Int64? = nil, modelOptions: ModelOptions) {
+    downloadStatus = status
+    if let code = errorCode {
+      self.errorCode = code
+    }
+    if let roughDuration = roughDownloadDuration {
+      roughDownloadDurationMs = roughDuration
+    }
+    if let exactDuration = exactDownloadDuration {
+      exactDownloadDurationMs = exactDuration
+    }
+    if let failureStatus = downloadFailureStatus {
+      self.downloadFailureStatus = failureStatus
+    }
+    options = modelOptions
+  }
+}
+
+/// Extension to build Firebase ML log event.
+extension FirebaseMlLogEvent {
+  mutating func setEvent(eventName: EventName, systemInfo: SystemInfo,
+                         modelDownloadLogEvent: ModelDownloadLogEvent) {
+    self.eventName = eventName
+    self.systemInfo = systemInfo
+    self.modelDownloadLogEvent = modelDownloadLogEvent
+  }
+}
+
+/// Data object for Firelog event.
+class FBMLDataObject: NSObject, GDTCOREventDataObject {
+  private let event: FirebaseMlLogEvent
+
+  init(event: FirebaseMlLogEvent) {
+    self.event = event
+  }
+
+  /// Encode Firelog event for transport.
+  func transportBytes() -> Data {
+    do {
+      // TODO: Should this be binary or json serialized?
+      let data = try event.serializedData()
```

Comment:
This is the data that will be sent to the server via Firelog. I'm a bit confused why we add `SwiftProtobuf` dependency if we serialize data with JSON. We either need to serialize data with protobuf here or remove any protobuf related code as not used. Could you please fix or add a TODO.

Reply:
So at this point, the event is serialized using SwiftProtobuf methods. It can be converted to binary or json (I'm using binary). Ultimately I'll rewrite this to be compatible with whatever replacement we end up using for SwiftProtobuf.

Reply:
Ah, sorry, my bad, I misread the code! All looks good here.